File: python/90_Subsets_II.py
#!/usr/bin/python

class Solution(object):
    def subsetsWithDup(self, nums):
    	n = len(nums)
    	nums = sorted(nums)
    	res = [[]]
    	part = []
    	def dfs(nums, part):
    		if len(nums) == 0:
    			return
    		i, last = 0, "*"
    		for i in range(len(nums)):
    			if nums[i] == last:
    				continue
    			# can't append part first
    			# since it would transfer address rather than value
    			# so once innter list changes, res change too
    			# part.append(nums[i])
    			res.append(part + [nums[i]]) 
    			dfs(nums[i + 1:], part + [nums[i]])
    			last = nums[i]
    	dfs(nums, part)
    	return res

def main():
    s = Solution()
    print s.subsetsWithDup([1,2,2])


if __name__ == "__main__":
    main()

File: c++/90_Subsets_II.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int> > subsetsWithDup(vector<int>& nums) {
        vector<vector<int> > res;
        vector<int> part;
        res.push_back(part);
        dfs(0, nums, part, res);
        return res;
    }

    void dfs(int start, vector<int> nums, vector<int> part, vector<vector<int> > &res) {
    	if (nums.size() == start)
    		return;
    	int last = 0;
    	for (int i = start; i < nums.size(); i++) {
    		if (i != start and nums[i] == last)
    			continue;
    		part.push_back(nums[i]);
    		res.push_back(part);
    		dfs(start + 1, nums, part, res);
    		part.pop_back();
    		last = nums[i];
    	}
    }
};


int main() {
	Solution s;
}